README.rst
============================================
TraitsUI: Traits-capable windowing framework
============================================

The TraitsUI project provides a toolkit-independent GUI abstraction layer,
which is used to support the "visualization" features of the
`Traits <http://github.com/enthought/traits>`__ package.
You can write a model using the Traits API and specify a GUI using the TraitsUI
API (views, items, editors, etc.), and let TraitsUI and your selected toolkit
back-end (Qt or Wx) take care of the details of displaying them.

Example
-------

Given a Traits model like the following::

    from traits.api import HasTraits, Str, Range, Enum

    class Person(HasTraits):
        name = Str('Jane Doe')
        age = Range(low=0)
        gender = Enum('female', 'male')

    person = Person(age=30)

And using TraitsUI to specify and display a GUI view::

    from traitsui.api import Item, RangeEditor, View

    person_view = View(
        Item('name'),
        Item('gender'),
        Item('age', editor=RangeEditor(mode='spinner', low=0, high=150)),
        buttons=['OK', 'Cancel'],
        resizable=True,
    )

    person.configure_traits(view=person_view)

It creates a GUI which looks like this:

.. image:: https://raw.github.com/enthought/traitsui/main/README_example.png

Important Links
---------------

- Website and Documentation: `<http://docs.enthought.com/traitsui>`__

  * User Manual `<http://docs.enthought.com/traitsui/traitsui_user_manual>`__
  * Tutorial `<http://docs.enthought.com/traitsui/tutorials>`__
  * API Documentation `<http://docs.enthought.com/traitsui/api>`__

- Source code repository: `<https://github.com/enthought/traitsui>`__

  * Issue tracker: `<https://github.com/enthought/traitsui/issues>`__

- Download releases: `<https://pypi.python.org/pypi/traitsui>`__

- Mailing list: `<https://groups.google.com/forum/#!forum/ets-users>`__

Installation
------------

If you want to run traitsui, you must also install:

- Traits `<https://github.com/enthought/traits>`__
- Pyface `<https://github.com/enthought/pyface>`__

You will also need one of the following backends:

- wxPython
- PySide2
- PyQt5

Backends have additional dependencies and there are optional dependencies on
NumPy and Pandas for some editors.

TraitsUI along with all dependencies can be installed in a straightforward way
using the `Enthought Deployment Manager <http://docs.enthought.com/edm/>`__,
``pip`` or other .

.. end_of_long_description

Running the Test Suite
----------------------

To run the test suite, you will need to install Git and
`EDM <http://docs.enthought.com/edm/>`__ as well as have a Python environment
which has install `Click <http://click.pocoo.org/>`__ available. You can then
follow the instructions in ``etstool.py``.  In particular::

    > python etstool.py test_all

will run tests in all supported environments automatically.


CHANGES.txt
Traits UI Changelog
===================

Release 8.0.0
-------------

This is a major release that provides support for PySide 6.4 and 6.5 as well
as Python 3.11.  It includes a number of backwards-incompatible changes
following the example of Pyface 8.0.0, most notable is moving the
`traitsui.qt4.*` modules to `traitsui.qt.*`.  As with Pyface 8.0 we include
a backwards-compatibility mode which allows importing from the 'qt4' namespace
via:

- setting the ETS toolkit to `"qt4"`
- setting the ETS_QT4_IMPORTS environment variable
- manually adding appropriate finders to sys.meta_path

This release also removes a number of features deprecated since TraitsUI 7.0.

Thanks To
---------

* David Baddeley
* Mark Dickinson
* Dominik Gresch
* hopeful0
* Chengyu Liu
* Orion Poplawski
* Corran Webster

Features
--------

* Be strict about 'handler.init()' return values (#2008)
* Move 'traitsui.qt4.*' to 'traitsui.qt.*' (#2004)
* Remove deprecated 'format' trait (#2002)
* Deprecate imports from traitsui.editors (use traitsui.editors.api) (#2000)
* Remove backwards-compatibility mode in undo/redo code (#1999)
* Support for Python 3.11 and PySide 6.4+ (#1994)

Fixes
-----

* Fix uses of None in date range editor (#2019)
* Add comments for Wx-only examples (#2011, #2020)
* Replace uses of "fast_ui" dispatch with "ui" dispatch (#2009)
* Fix FileDialog selection actions (#2003)
* Fix regression in Wx version of FileEditor (#1993)
* Fix missing attribute of InstanceFactoryChocie (#1989)

Test suite
----------

* Replace some relative imports in tests (#1985)

Release 7.4.3
-------------

This is a small bugfix release which resolves some bugs that have come to light
as we get more experience with PySide 6, together with changes to run CI on
more recent Python versions and current Github infrastructure.

Thanks To
---------

* Mark Dickinson
* Matt Reay
* Corran Webster
* John Wiggins

Fixes
-----
* Fix rendering of dragged tree nodes on Qt6 (#1960)
* Add explicit allow_none in Datetime traits (#1964)
* Generate extra TableEditor menus dynamically (#1966)
* Display the selected date in the DateEditor in custom style (#1967)
* Guard Qt TreeEditor against destroyed QTreeViewItems (#1973)

Test suite
----------
* Update CI for EDM Python 3.8 (#1972)

Release 7.4.2
-------------

This is a small bug fix release which resolves a couple of bugs in the
RangeEditor and TreeEditor.  It also pins the PySide6 version to < 6.4.0 due
to incompatibilities with the new enum system, which should improve the install
experience.

Thanks To
---------

* Aaron Ayres
* hopeful0
* Corran Webster

Fixes
-----

* Fix RangeTextEditor handling of None for low/high (#1950)
* Pin PySide6 to < 6.4 (#1951)
* Fix issues with shared "New" context menu in TreeEditor (#1953)

Release 7.4.1
-------------

This is a bug fix release which resolves an issue with RangeEditor not handling
None for either the high or low value, as well as an issue with ProgressColumn
rendering.

Thanks To
---------

* Steve Allen
* Chris Angell
* Aaron Ayres
* Maxwell Grady
* Johannes Loibl
* Rahul Poruri
* Corran Webster
* Hai Yan

Fixes
-----
* Fix issue with progress column rendering on Linux and Windows. (#1937)
* Fix typos in docstrings (#1935)
* Fix range text editor bug (#1931)
* Fix issue where file editor filters were not used in simple editor (#1930)

Release 7.4.0
-------------

This is a minor release which fixes a number of bugs and adds a couple of
small features.  The most significant changes are the ability to add separators
to EnumEditor comboboxes, the ability to use Pyface action Schemas for menu bars
and toolbars in Views, and getting the VideoEditor working against the new
QtMultimedia APIs in Qt6.

Thanks To
~~~~~~~~~

* Aaron Ayres
* Mark Dickinson
* Rahul Poruri
* PyHannes
* Prabhu Ramachandran
* Ioannis Tziakos
* Corran Webster
* John Wiggins
* Hai Yan

Features
~~~~~~~~
* Separators in Combobox EnumEditors. (#1885)
* Allow the use of Pyface Schemas for View menus and toolbars. (#1827)

Fixes
~~~~~
* Delay imports which force toolkit selection (#1883)
* Fix ImageEditor paintEvent when image is None. (#1907)
* Fix VideoEditor for Qt6 (#1908)
* Fix the ImageEnumEditor on Qt5+ (#1910)
* Fix some height and width calls for Qt and Python 3.10+ (#1911)
* Fix a crash on PyQt5 when a Group has no content (#1914)
* Numerous fixes for tests and CI (#1897, #1893, #1889, #1894, #1898, #1899,
  #1990, #1902, #1903)

Documentation changes
~~~~~~~~~~~~~~~~~~~~~
* Add a copy button to code blocks in documentation (#1904)

Release 7.3.1
-------------

TraitsUI 7.3.1 is a bugfix release that resolves a couple of critical errors in
some Editors.

Thanks To
~~~~~~~~~

* Mark Dickinson
* Steven Kern
* Orion Poplawski
* Rahul Poruri
* Corran Webster

Fixes
~~~~~


etstool.py
# (C) Copyright 2004-2023 Enthought, Inc., Austin, TX
# All rights reserved.
#
# This software is provided without warranty under the terms of the BSD
# license included in LICENSE.txt and may be redistributed only under
# the conditions described in the aforementioned license. The license
# is also available online at http://www.enthought.com/licenses/BSD.txt
#
# Thanks for using Enthought open source!

"""
Tasks for Test Runs
===================

This file is intended to be used with a python environment with the
click library to automate the process of setting up test environments
and running the test within them.  This improves repeatability and
reliability of tests be removing many of the variables around the
developer's particular Python environment.  Test environment setup and
package management is performed using `EDM
<http://docs.enthought.com/edm/>`_

To use this to run you tests, you will need to install EDM and click
into your working environment.  You will also need to have git
installed to access required source code from github repositories.
You can then do::

    python etstool.py install --runtime=... --toolkit=...

to create a test environment from the current codebase and::

    python etstool.py test --runtime=... --toolkit=...

to run tests in that environment.  You can remove the environment with::

    python etstool.py cleanup --runtime=... --toolkit=...

If you need to make frequent changes to the source, it is often convenient
to install the source in editable mode::

    python etstool.py install --editable --runtime=... --toolkit=...

You can run all three tasks at once with::

    python etstool.py test_clean --runtime=... --toolkit=...

which will create, install, run tests, and then clean-up the environment.  And
you can run tests in all supported runtimes and toolkits (with cleanup)
using::

    python etstool.py test_all

Currently supported runtime values are ``3.6``, and currently
supported toolkits are ``null``, ``pyqt5``, ``pyside2`` and ``wx``.
Not all combinations of toolkits and runtimes will work, but the tasks will
fail with a clear error if that is the case.

Tests can still be run via the usual means in other environments if that suits
a developer's purpose.

Tasks for generating documentation
==================================

First, install or update the development environment::

    python etstool.py install --runtime=... --toolkit=...

Then, run the command to build documentation::

    python etstool.py docs --runtime=... --toolkit=...

Changing This File
------------------

To change the packages installed during a test run, change the dependencies
variable below.  To install a package from github, or one which is not yet
available via EDM, add it to the `ci-src-requirements.txt` file (these will be
installed by `pip`).

Other changes to commands should be a straightforward change to the listed
commands for each task. See the EDM documentation for more information about
how to run commands within an EDM enviornment.

Build changelog
---------------
To create a first-cut changelog from the news fragments, use this command::

    python etstool.py changelog build

This will update the changelog file. You should review and edit it.
"""

import glob
import os
import subprocess
import sys
from contextlib import contextmanager
from shutil import copy as copyfile
from shutil import rmtree
from tempfile import mkdtemp

import click


EDM_CONFIG_LOCATION = os.path.abspath(os.path.join(
    os.path.dirname(__file__),
    'edm.yaml',
))

supported_combinations = {
    '3.8': {'pyside2', 'pyside6', 'pyqt5', 'pyqt6', 'wx', 'null'},
}

# Default Python version to use in the comamnds below if none is specified.
DEFAULT_RUNTIME = '3.6'

# Default toolkit to use if none specified.
DEFAULT_TOOLKIT = 'null'

# Required runtime dependencies. Should match pyproject.toml.
dependencies = {
    "pyface>=8.0",
    "traits",
    "setuptools",
}

# Dependencies we install from source for cron tests
# Order from packages with the most dependencies to one with the least
# dependencies. Packages are forced re-installed in this order.
source_dependencies = {
    "pyface": "git+http://github.com/enthought/pyface.git#egg=pyface",
    "traits": "git+http://github.com/enthought/traits.git#egg=traits",
}

# The following should match extras_require in pyproject.toml but with package
# names compatible with EDM
extra_dependencies = {
    'pyside2': {
        'pygments',
    },
    'pyside6': {
        'pygments',
    },
    'pyqt5': {
        'pygments',
    },
    'pyqt6': {
        'pygments',
    },
    # XXX once wxPython 4 is available in EDM, we will want it here
    'wx': {
        'numpy',
    },
    'null': set(),
    # Toolkit-independent dependencies for demo testing.
    'examples': {
        'apptools',
        'chaco',
        'h5py',
        'numpy',
        'pandas',
    },
    # Optional dependencies for some editors
    'editors': {
        'numpy',
        'pandas',
    },
    # Test dependencies also applied to installation from PYPI
    'test': {
        'packaging',
    },
}

# Extra runtime dependencies
runtime_dependencies = {
    "3.8": {"pillow_simd", "tables"},
}

# Dependencies for CI jobs using this file.
ci_dependencies = {
    "coverage",
    "flake8",
    "flake8_ets",
}

doc_dependencies = {
    "sphinx",
    "enthought_sphinx_theme",
    "pyface",
}

#: Paths to ignore in Sphinx-apidoc
doc_ignore = {
    "traitsui/qt/*",
    "traitsui/wx/*",
    "*/tests",
}

environment_vars = {
    'pyside2': {'ETS_TOOLKIT': 'qt4', 'QT_API': 'pyside2'},
    'pyside6': {'ETS_TOOLKIT': 'qt4', 'QT_API': 'pyside6'},
    'pyqt5': {"ETS_TOOLKIT": "qt4", "QT_API": "pyqt5"},
    'pyqt6': {"ETS_TOOLKIT": "qt4", "QT_API": "pyqt6"},
    'wx': {'ETS_TOOLKIT': 'wx'},
    'null': {'ETS_TOOLKIT': 'null'},
}

# toolkit versions in EDM
edm_versions = {
    ('3.8', 'pyside6'),
    ('3.8', 'pyqt6'),
}

# Location of documentation files
HERE = os.path.dirname(__file__)
DOCS_DIR = os.path.join(HERE, "docs")

# Location of news fragment for creating changelog.
NEWS_FRAGMENT_DIR = os.path.join(DOCS_DIR, "releases", "upcoming")

# Location of the Changelog file.
CHANGELOG_PATH = os.path.join(HERE, "CHANGES.txt")


def normalize(name):
    return name.replace("_", "-")


# Ensure that "-h" is supported for getting help.
CONTEXT_SETTINGS = dict(
    help_option_names=["-h", "--help"],
    token_normalize_func=normalize,
)


@click.group(context_settings=CONTEXT_SETTINGS)
def cli():
    pass


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
@click.option('--environment', default=None)
@click.option(
    "--editable/--not-editable",
    default=False,
    help="Install main package in 'editable' mode?  [default: --not-editable]",
)
@click.option('--source/--no-source', default=False)
def install(runtime, toolkit, environment, editable, source):
    """Install project and dependencies into a clean EDM environment."""
    parameters = get_parameters(runtime, toolkit, environment)
    packages = (
        dependencies
        | extra_dependencies.get(toolkit, set())
        | extra_dependencies["test"]
        | extra_dependencies["examples"]
        | extra_dependencies["editors"]
        | runtime_dependencies[runtime]
        | ci_dependencies
    )

    # edm commands to setup the development environment
    commands = [
        "edm -c {config} environments create {environment} --force --version={runtime}",
        "edm -c {config} install -y -e {environment} " + " ".join(packages),
    ]

    if (runtime, toolkit) in edm_versions:
        commands.append(
            "edm -c {config} install -y -e {environment} {toolkit}"
        )
    elif toolkit == 'wx':
        if sys.platform != 'linux':
            commands.append("edm -c {config} run -e {environment} -- pip install wxPython")
        else:
            # XXX this is mainly for TravisCI workers; need a generic solution
            commands.append(
                "edm -c {config} run -e {environment} -- pip install -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-16.04/ wxPython"
            )
    elif toolkit == 'pyside6':
        commands.append('edm run -e {environment} -- pip install "pyside6"')
    elif toolkit != "null":
        commands.append("edm -c {config} run -e {environment} -- pip install {toolkit}")

    commands.extend(
        [
            "edm -c {config} run -e {environment} -- pip install --force-reinstall -r ci-src-requirements.txt --no-dependencies",
        ]
    )

    click.echo("Creating environment '{environment}'".format(**parameters))
    execute(commands, parameters)

    if source:
        cmd_fmt = (
            "edm -c {config} plumbing remove-package --environment {environment} --force "
        )
        commands = [
            cmd_fmt + dependency for dependency in source_dependencies.keys()
        ]
        execute(commands, parameters)
        source_pkgs = source_dependencies.values()
        # Without the --no-dependencies flag such that new dependencies on
        # main branch are brought in.
        commands = [
            "python -m pip install --force-reinstall {pkg}".format(pkg=pkg)
            for pkg in source_pkgs
        ]
        commands = [
            "edm -c {config} run -e {environment} -- " + command for command in commands
        ]
        execute(commands, parameters)

    # Always install local source again with no dependencies
    # to mitigate risk of testing against a distributed release.
    install_traitsui = (
        "edm -c {config} run -e {environment} -- "
        "pip install --force-reinstall  --no-dependencies "
    )
    if editable:
        install_traitsui += "--editable "
    install_traitsui += "."
    execute([install_traitsui], parameters)

    click.echo('Done install')


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
@click.option('--environment', default=None)
def shell(runtime, toolkit, environment):
    """Create a shell into the EDM development environment
    (aka 'activate' it).

    """
    parameters = get_parameters(runtime, toolkit, environment)
    commands = [
        "edm -c {config} shell -e {environment}",
    ]
    execute(commands, parameters)


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
@click.option('--environment', default=None)
def test(runtime, toolkit, environment):
    """Run the test suite in a given environment with the specified toolkit."""
    parameters = get_parameters(runtime, toolkit, environment)
    environ = environment_vars.get(toolkit, {}).copy()
    environ['PYTHONUNBUFFERED'] = "1"

    parameters["integrationtests"] = os.path.abspath("integrationtests")
    commands = [
        "edm -c {config} run -e {environment} -- python -X faulthandler -W default -m coverage run -p -m unittest discover -v traitsui",
        # coverage run prevents local images to be loaded for demo examples
        # which are not defined in Python packages. Run with python directly
        # instead.
        "edm -c {config} run -e {environment} -- python -X faulthandler -W default -m unittest discover -v {integrationtests}",
    ]

    # We run in a tempdir to avoid accidentally picking up wrong traitsui
    # code from a local dir.  We need to ensure a good .coveragerc is in
    # that directory, plus coverage has a bug that means a non-local coverage
    # file doesn't get populated correctly.
    click.echo("Running tests in '{environment}'".format(**parameters))
    with do_in_tempdir(files=['.coveragerc'], capture_files=['./.coverage*']):
        os.environ.update(environ)
        execute(commands, parameters)
    click.echo('Done test')


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
@click.option('--environment', default=None)
def cleanup(runtime, toolkit, environment):
    """Remove a development environment."""
    parameters = get_parameters(runtime, toolkit, environment)
    commands = [
        "edm -c {config} environments remove {environment} --purge -y",
    ]
    click.echo("Cleaning up environment '{environment}'".format(**parameters))
    execute(commands, parameters)
    click.echo('Done cleanup')


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
def test_clean(runtime, toolkit):
    """Run tests in a clean environment, cleaning up afterwards"""
    args = ['--toolkit={}'.format(toolkit), '--runtime={}'.format(runtime)]
    try:
        install(args=args, standalone_mode=False)
        test(args=args, standalone_mode=False)
    finally:
        cleanup(args=args, standalone_mode=False)


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
@click.option('--environment', default=None)
def docs(runtime, toolkit, environment):
    """Autogenerate documentation"""
    parameters = get_parameters(runtime, toolkit, environment)
    packages = ' '.join(doc_dependencies)
    ignore = " ".join(doc_ignore)
    commands = [
        "edm -c {config} install -y -e {environment} " + packages,
    ]
    click.echo(
        "Installing documentation tools in  '{environment}'".format(
            **parameters
        )
    )
    execute(commands, parameters)
    click.echo('Done installing documentation tools')

    click.echo(
        "Regenerating API docs in  '{environment}'".format(**parameters)
    )
    output_path = os.path.join("docs", "source", "api")
    if not os.path.exists(output_path):
        os.makedirs(output_path)
    commands = [
        "edm -c {config} run -e {environment} -- sphinx-apidoc "
        "-e "
        "-M "
        "--no-toc "
        "-o " + output_path + " traitsui " + ignore
    ]
    execute(commands, parameters)
    click.echo("Done regenerating API docs")

    os.chdir('docs')
    command = (
        "edm -c {config} run -e {environment} -- sphinx-build -b html "
        "-d build/doctrees "
        "source "
        "build/html"
    )
    click.echo(
        "Building documentation in  '{environment}'".format(**parameters)
    )
    try:
        execute([command], parameters)
    finally:
        os.chdir('..')
    click.echo('Done building documentation')


@cli.command()
def test_all():
    """Run test_clean across all supported environment combinations."""
    failed_command = False
    for runtime, toolkits in supported_combinations.items():
        for toolkit in toolkits:
            args = [
                '--toolkit={}'.format(toolkit),
                '--runtime={}'.format(runtime),
            ]
            try:
                test_clean(args, standalone_mode=True)
            except SystemExit:
                failed_command = True
    if failed_command:
        sys.exit(1)


@cli.command()
@click.option('--runtime', default=DEFAULT_RUNTIME)
@click.option('--toolkit', default=DEFAULT_TOOLKIT)
@click.option(
    "--environment", default=None, help="Name of EDM environment to check."
)
@click.option(
    '--strict/--not-strict',
    default=False,
    help="Use strict configuration for flake8 [default: --not-strict]",
)
def flake8(runtime, toolkit, environment, strict):
    """Run a flake8 check in a given environment."""
    parameters = get_parameters(runtime, toolkit, environment)
    config = ""
    if strict:
        config = "--config=flake8_strict.cfg "
    commands = [
        "edm -c {config} run -e {environment} -- python -m flake8 " + config,
    ]
    execute(commands, parameters)


@cli.group("changelog")
@click.pass_context
def changelog(ctx):
    """Group of commands related to creating changelog."""

    ctx.obj = {
        # Mapping from news fragment type to their description in
        # the changelog.
        "type_to_description": {
            "feature": "Features",
            "bugfix": "Fixes",
            "deprecation": "Deprecations",
            "removal": "Removals",
            "doc": "Documentation changes",
            "test": "Test suite",
            "build": "Build System",
        }
    }


@changelog.command("create")
@click.pass_context
def create_news_fragment(ctx):
    """Create a news fragment for your PR."""

    pr_number = click.prompt('Please enter the PR number', type=int)
    type_ = click.prompt(
        "Choose a fragment type:",
        type=click.Choice(ctx.obj["type_to_description"]),
    )

    filepath = os.path.join(NEWS_FRAGMENT_DIR, f"{pr_number}.{type_}.rst")

    if os.path.exists(filepath):
        click.echo("FAILED: File {} already exists.".format(filepath))
        ctx.exit(1)

    content = click.prompt(
        "Describe the changes to the END USERS.\n"
        "Example: 'Remove subpackage xyz.'\n",
        type=str,
    )
    if not os.path.exists(NEWS_FRAGMENT_DIR):
        os.makedirs(NEWS_FRAGMENT_DIR)
    with open(filepath, "w", encoding="utf-8") as fp:
        fp.write(content + f" (#{pr_number})")

    click.echo("Please commit the file created at: {}".format(filepath))


@changelog.command("build")
@click.pass_context
def build_changelog(ctx):
    """Build Changelog created from all the news fragments."""
    # This is a rather simple first-cut generation of the changelog.
    # It removes the laborious concatenation, but the end results might
    # still require some tweaking.
    contents = []

    # Collect news fragment files as we go, and then optionally remove them.
    handled_file_paths = []

    for type_, description in ctx.obj["type_to_description"].items():
        pattern = os.path.join(NEWS_FRAGMENT_DIR, f"*.{type_}.rst")
        file_paths = sorted(glob.glob(pattern))

        if file_paths:
            contents.append("")
            contents.append(description)
            contents.append("-" * len(description))

        for filename in file_paths:
            with open(filename, "r", encoding="utf-8") as fp:
                contents.append("* " + fp.read())
            handled_file_paths.append(filename)

    # Prepend content to the changelog file.

    with open(CHANGELOG_PATH, "r", encoding="utf-8") as fp:
        original_changelog = fp.read()

    with open(CHANGELOG_PATH, "w", encoding="utf-8") as fp:
        if contents:
            print(*contents, sep="\n", file=fp)
        fp.write(original_changelog)

    click.echo(f"Changelog is updated. Please review it at {CHANGELOG_PATH}")

    # Optionally clean up collected news fragments.
    should_clean = click.confirm("Do you want to remove the news fragments?")
    if should_clean:
        for file_path in handled_file_paths:
            os.remove(file_path)

        # Report any leftover for developers to inspect.
        leftovers = sorted(glob.glob(os.path.join(NEWS_FRAGMENT_DIR, "*")))
        if leftovers:
            click.echo("These files are not collected:")
            click.echo("\n  ".join([""] + leftovers))

    click.echo("Done")


# ----------------------------------------------------------------------------
# Utility routines
# ----------------------------------------------------------------------------


def get_parameters(runtime, toolkit, environment, config=EDM_CONFIG_LOCATION):
    """Set up parameters dictionary for format() substitution"""
    parameters = {
        'runtime': runtime,
        'toolkit': toolkit,
        'environment': environment,
        'config': config,
    }
    if toolkit not in supported_combinations[runtime]:
        msg = (
            "Python {runtime} and toolkit {toolkit} not supported by "
            + "test environments"
        )
        raise RuntimeError(msg.format(**parameters))
    if environment is None:
        parameters['environment'] = 'traitsui-test-{runtime}-{toolkit}'.format(
            **parameters
        )
    return parameters


@contextmanager
def do_in_tempdir(files=(), capture_files=()):
    """Create a temporary directory, cleaning up after done.

    Creates the temporary directory, and changes into it.  On exit returns to
    original directory and removes temporary dir.

    Parameters
    ----------
    files : sequence of filenames
        Files to be copied across to temporary directory.
    capture_files : sequence of filenames
        Files to be copied back from temporary directory.
    """
    path = mkdtemp()
    old_path = os.getcwd()

    # send across any files we need
    for filepath in files:
        click.echo('copying file to tempdir: {}'.format(filepath))
        copyfile(filepath, path)

    os.chdir(path)
    try:
        yield path
        # retrieve any result files we want
        for pattern in capture_files:
            for filepath in glob.iglob(pattern):
                click.echo('copying file back: {}'.format(filepath))
                copyfile(filepath, old_path)
    finally:
        os.chdir(old_path)
        rmtree(path)


def execute(commands, parameters):
    for command in commands:
        click.echo("[EXECUTING] {}".format(command.format(**parameters)))
        try:
            subprocess.check_call(
                [arg.format(**parameters) for arg in command.split()]
            )
        except subprocess.CalledProcessError as exc:
            click.echo(str(exc))
            sys.exit(1)


if __name__ == '__main__':
    cli(prog_name="python etstool.py")
